### DISCUSSION TITLE
Is there a way to return the same object from Rust side (singleton support)?

### DISCUSSION THREAD

**Discussion by lycantropos** [PyO3/pyo3#2382] [Questions]
Imagine I have some `Rust` enumeration that I want to export to Python
```rust
#[pyclass(name = "MyEnum", module = "my_enum")]
#[derive(Clone)]
enum MyEnum {
 Value,
}
```
and I also want it to be constructible in Python
```rust
#[pymethods]
impl MyEnum {
    #[new]
    fn __new__() -> MyEnum {
        MyEnum::Value
    }
}
```
finally module definition could be simply
```rust
#[pymodule]
fn my_module(_py: Python, module: &PyModule) -> PyResult<()> {
    module.add_class::<MyEnum>()?;
    Ok(())
}
```

The problem is that I want to return the same object so that Python's `is` operator works (like for Python's `enum.Enum`s), but I get
```python
>>> from my_module import MyEnum
>>> MyEnum() is MyEnum.Value
False
```
is there a way to do that?

**Comment by davidhewitt** [comment]
Maybe something like this?

```rust
use pyo3::prelude::*;

#[pyclass]
struct Singleton { }

#[pymethods]
impl Singleton {
    #[classattr]
    #[allow(non_snake_case)]
    fn VALUE() -> Py<Self> {
        Python::with_gil(|py| Py::new(py, Singleton { })).unwrap()
    }
}
```

I don't think it's possible for `#[new]` in PyO3 to return an existing object (at the moment), though it really should be.

**Reply by lycantropos** [reply]
I'm able to create singleton'ish class attributes by using `const`, your suggestion works too, but in my particular case I was thinking about a function that given an integer will return one of possible variants so they have the same address on each invocation

**Reply by lycantropos** [reply]
because now it seems to be making `clone`s each time called

**Reply by davidhewitt** [reply]
I think you might be able to do this with a `#[staticmethod]` which looks up existing attributes. 

Or alternatively, are you actually highlighting that there's a bug with `#[pyclass] enum` in that it doesn't return singletons?

**Reply by lycantropos** [reply]
it returns the same object when I use `#[classattr]`, but when I try to return `#[classattr]` attribute from method it creates a new instance, here is my real case scenario:
https://github.com/lycantropos/rene/blob/ed3b09ddaebc79556cacd7cbc8ba0dd6835971ff/src/lib.rs#L40-L91

my end goal is to be able to construct enum variants with `int` value from Python's side so that it returns same Python objects, currently it creates new instances on each call, so after building project locally (e.g. with `python setup.py build_ext --inplace`) I'm getting
```python
>>> from rene import Orientation
>>> Orintation(0) is Orientation.COLLINEAR
False
```
while it works for Python's `enum.Enum` classes

**Reply by davidhewitt** [reply]
Right, sorry for the long delay on this. I think you can achieve something like this by replacing

```rust
return Ok(PyOrientation::CLOCKWISE)
```

with

```rust
return PyOrientation::type_object(py).getattr("CLOCKWISE")
```

i.e. you need to return the already-constructed instance on the Python type object. There's probably something that can be done on PyO3's side to make this nicer, but it's not immediately obvious what.

**Reply by lycantropos** [reply]
I'm still getting a new object instead of already constructed one, `is` check returns `False`
